build-system/pr-check.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const BBPromise = require('bluebird');
+const child_process = require('child_process');
+const exec = BBPromise.promisify(child_process.exec);
+const fs = BBPromise.promisifyAll(require('fs'));
+const path = require('path');
+
+/**
+ * @param {string} str
+ * @param {string? prefix
+ * @return {boolean}
+ */
+function hasPrefix(str, prefix) {
+  return str.indexOf(prefix) == 0;
+}
+
+/**
+ * @param {string} str
+ * @param {string? prefix
+ * @return {boolean}
+ */
+function hasSuffix(str, suffix) {
+  return str.lastIndexOf(suffix) == str.length - suffix.length;
+}
+
+/**
+ * @param {string} travisCommitRange
+ * @return {Promise<!Array<string>>}
+ */
+function filesInPr(travisCommitRange) {
+  return exec(`git diff --name-only ${travisCommitRange}`)
+      .then(output => output.trim().split('\n'));
+}
+
+/**
+ * @param {string} filePath
+ * @return {boolean}
+ */
+function isValidatorWebuiFile(filePath) {
+  return hasPrefix(filePath, 'validator/webui');
+}
+
+/**
+ * @param {string} filePath
+ * @return {boolean}
+ */
+function isValidatorFile(filePath) {
+  if (hasPrefix(filePath, 'validator/')) return true;
+  if (!hasSuffix(path.dirname(filePath), '0.1') &&
+      !hasSuffix(path.dirname(filePath), 'test'))
+    return false;
+  const name = path.basename(filePath);
+  return hasPrefix(name, 'validator-') &&
+      (hasSuffix(name, '.out') || hasSuffix(name, '.html') ||
+       hasSuffix('.protoascii'));
+}
+
+/**
+ * @param {!Array<string>} filePaths
+ * @returns {!Set<string>}
+ */
+function determineBuildTargets(filePaths) {
+  const targetSet = new Set();
+  for (p of filePaths) {
+    if (isValidatorWebuiFile(p)) {
+      targetSet.add('VALIDATOR_WEBUI');
+    } else if (isValidatorFile(p)) {
+      targetSet.add('VALIDATOR');
+    } else {
+      targetSet.add('RUNTIME');
+    }
+  }
+  return targetSet;
+}
+
+if (process.argv.length <= 2) {
+  console.log(`Usage: ${__filename} TRAVIS_COMMIT_RANGE`);
+  process.exit(-1);
+}
+const travisCommitRange = process.argv[2];
+filesInPr(travisCommitRange).then(determineBuildTargets).then(buildTargets => {
+  for (const target of buildTargets) {
+    console.log('target detected: ' + target);
+  }
+  const steps = [];
+  console.log('scheduling COMMON steps');
+  steps.push(exec('npm run ava'));
```

Comment:
this is probably the reason why its erroring out right now, the error looks like dist executed before build had finished.

also its bad that even though are errors the ci is green, we need to make sure it actually fails the travis run.